tasks/rescueStuckScrollTxn.ts

```diff
@@ -0,0 +1,146 @@
+        name: "messageNonce",
+        type: "uint256",
+      },
+      {
+        indexed: false,
+        internalType: "uint256",
+        name: "gasLimit",
+        type: "uint256",
+      },
+      {
+        indexed: false,
+        internalType: "bytes",
+        name: "message",
+        type: "bytes",
+      },
+    ],
+    name: "SentMessage",
+    type: "event",
+  },
+  {
+    inputs: [
+      {
+        internalType: "address",
+        name: "_from",
+        type: "address",
+      },
+      {
+        internalType: "address",
+        name: "_to",
+        type: "address",
+      },
+      {
+        internalType: "uint256",
+        name: "_value",
+        type: "uint256",
+      },
+      {
+        internalType: "uint256",
+        name: "_messageNonce",
+        type: "uint256",
+      },
+      {
+        internalType: "bytes",
+        name: "_message",
+        type: "bytes",
+      },
+      {
+        internalType: "uint32",
+        name: "_newGasLimit",
+        type: "uint32",
+      },
+      {
+        internalType: "address",
+        name: "_refundAddress",
+        type: "address",
+      },
+    ],
+    name: "replayMessage",
+    outputs: [],
+    stateMutability: "payable",
+    type: "function",
+  },
+];
+
+task("rescue-stuck-scroll-txn", "Rescue a failed Scroll transaction")
+  .addParam("l1Hash", "Txn of the L1 message to rescue")
+  .addParam("gasLimit", "Gas limit to use for the rescue transaction")
+  .setAction(async function (taskArguments, hre_: any) {
+    const chainId = await hre_.getChainId();
+    if (!["1", "11155111"].includes(String(chainId))) {
+      throw new Error("This script can only be run on Sepolia or Ethereum mainnet");
+    }
+    const signer = (await hre_.ethers.getSigners())[0] as unknown as Signer;
+    const messengerContract = new Contract(L1_ADDRESS_MAP[chainId].scrollMessengerRelay, relayMessengerAbi, signer);
+
+    const txn = await signer.provider?.getTransactionReceipt(taskArguments.l1Hash);
+    const relevantEvent = txn?.logs?.find(
+      (log) => log.topics[0] === messengerContract.interface.getEventTopic("SentMessage")
+    );
+    if (!relevantEvent) {
+      throw new Error("No relevant event found. Is this a Scroll bridge transaction?");
+    }
+    const decodedEvent = messengerContract.interface.parseLog(relevantEvent);
+    const { sender, target, value, messageNonce, message } = decodedEvent.args;
+    const refundAddress = await signer.getAddress();
+
+    console.debug("Log found. Event Decoded.");
+    console.debug("Will replay with these parameters:", {
+      _from: sender,
+      _to: target,
+      _value: value.toString(),
+      _messageNonce: messageNonce.toString(),
+      _message: message.toString(),
+      _newGasLimit: taskArguments.gasLimit,
+      _refundAddress: refundAddress,
+    });
+    console.debug("Replaying message (sending with 0.001ETH )...");
+    const resultingTxn = await messengerContract.replayMessage(
```

Comment:
Wait, if an L2 message fails because of gas, we have to re-send it from _L1_? There's no way to replay it on the L2, right?

Reply:
Per the TG chat we have to replay via an L1 transaction. The funds being rescued are on L2

This is a similar process to how the Scroll Bridge UI performs retry attempts for failed bridge attempts

Reply:
Does that mean we can "replay" this message by sending an EOA txn or do we need to get the HubPool to resend it?

Reply:
Correct - this function is permissionless. The input values to this function are enriched with the internal state of the Scroll messenger before being submitted to the queue. 

The L2 layer will also ensure that only one successful message is transmitted from L1 -> L2. Technically, you could call a "rescue" on successful relay attempts, but that would fail when it landed in Scroll